cli/azd/cmd/env.go

```diff
@@ -374,17 +378,47 @@ func newEnvRefreshAction(
 		userProfileService:         userProfileService,
 		subscriptionTenantResolver: subscriptionTenantResolver,
 		alphaFeatureManager:        alphaFeatureManager,
+		serviceLocator:             serviceLocator,
 	}
 }
 
 func (ef *envRefreshAction) Run(ctx context.Context) (*actions.ActionResult, error) {
-	if err := ef.projectManager.Initialize(ctx, ef.projectConfig); err != nil {
+	// We want to support the usual -e / --environment arguments as all our commands which take environments do, but for
+	// ergonomics, we'd also like you to be able to run `azd env refresh some-environment-name` to behave the same way as
+	// `azd env refresh -e some-environment-name` would have.
+
+	if len(ef.args) == 1 && ef.flags.envFlag.environmentName != "" && ef.args[0] != ef.flags.envFlag.environmentName {
+		return nil, errors.New(
+			"the --environment flag and an explicit environment name as an argument may not be used together")
+	}
+
+	if len(ef.args) == 1 {
+		ef.flags.envFlag.environmentName = ef.args[0]
+	}
+
+	env, err := loadOrCreateEnvironment(ctx, ef.flags.envFlag.environmentName, ef.azdCtx, ef.console)
+	if err != nil {
+		return nil, fmt.Errorf("loading environment: %w", err)
+	}
+
+	// Now that we've selected the right environment, we can create our dependencies that require an environment.
+	projectManager := project.NewProjectManager(ef.azdCtx,
```

Comment:
I couldn't really figure out a better way to do this.  The problem is that these two components (the service manager and the resource manager) need the environment, but we don't know what environment we want to load until we've done some work in this command, so I can't leverage the auto injecting stuff from the IOC container.

I played around with trying to use `lazy.Lazy` here, but it didn't seem to work out great.

Any thoughts on how to improve this?

Reply:
@ellismg Ideally, it should be possible to affect the lifecycle prior to "action" construction, where we can read flags/args as "config", and construct the action based on the config.

Reply:
IDK if it makes sense, but does it work if we define an `Args` [function ](https://github.com/spf13/cobra/blob/main/user_guide.md) that validates + sets `envFlag` value, and the happy path works as before? The only thing that could be affected is usage telemetry, but it only examines the `cobra` layer which works the same.

Reply:
> IDK if it makes sense, but does it work if we define an Args [function ](https://github.com/spf13/cobra/blob/main/user_guide.md)that validates + sets envFlag value, and the happy path works as before? The only thing that could be affected is usage telemetry, but it only examines the cobra layer which works the same.

Ooh, I think that may work, I will take a whirl with that.

Reply:
Indeed - it did work, and it is very clean!